src/components/Article.jsx

```diff
@@ -1,21 +1,40 @@
 import { useState, useEffect } from "react";
-import { getArticle, getComments } from "../api";
-import { useLocation } from "react-router-dom";
+import { getArticle, getComments, updateVotes } from "../api";
+import { useParams } from "react-router-dom";
 import CommentCard from "./CommentCard";
+import ThumbUpOutlinedIcon from "@mui/icons-material/ThumbUpOutlined";
+import ThumbDownOutlinedIcon from "@mui/icons-material/ThumbDownOutlined";
 
 const Article = (props) => {
   const [fullArticle, setFullArticle] = useState({});
   const [comments, setComments] = useState([]);
-  const location = useLocation();
-  const { article_id } = location.state;
+  const [vote, setVote] = useState(0);
+  const [voteError, setVoteError] = useState("");
+  const { article_id } = useParams();
 
   useEffect(() => {
-    getArticle(article_id).then((result) => setFullArticle(result.article));
-  }, []);
+    getArticle(article_id).then((result) => {
+      setFullArticle(result.article);
+      setVote(result.article.votes);
+    });
+  }, [article_id]);
+
   useEffect(() => {
     getComments(article_id).then((result) => setComments(result.comments));
   }, []);
 
+  const handleUpVote = () => {
+    setVote(vote + 1);
+    updateVotes(article_id, 1).catch((error) => {
+      setVoteError(`Vote Error: ${error}`);
+    });
+  };
+  const handleDownVote = () => {
+    setVote(vote - 1);
+    updateVotes(article_id, -1).catch((error) => {
+      setVoteError(`Vote Error: ${error}`);
+    });
+  };
```

Comment:
Some good implementation of optimistic rendering - the main improvement I suggest is correcting the vote's state when an error occurs, so if a downvote request has gone wrong, increase the votes by 1, and vice versa.
Great that you have some visual indication for the user that something has gone wrong though!

If you wanted a challenge - you could make it so the user is limited to either one upvote or one downvote 👀